op-e2e/actions/proofs/holocene_test.go

```diff
@@ -0,0 +1,193 @@
+	blockModifiers      []actionsHelpers.BlockModifier
+	frames              []uint
+	safeHeadPreHolocene uint64
+	safeHeadHolocene    uint64
+}
+
+// blockFudger invalidates the signature for the second transaction in the block.
+// This should result in an invalid payload in the engine queue.
+var blockFudger = func(block *types.Block) *types.Block {
+	alice := types.NewCancunSigner(big.NewInt(901))
+	txs := block.Transactions()
+	newTx, err := txs[1].WithSignature(alice, make([]byte, 65))
+	newTx.IsDepositTx()
+	if err != nil {
+		panic(err)
+	}
+	txs[1] = newTx
+	return block
+}
+
+// blockSpudger invalidates the parentHash of the block
+var blockSpudger = func(block *types.Block) *types.Block {
+	headerCopy := block.Header()
+	headerCopy.ParentHash = common.MaxHash
+	return block.WithSeal(headerCopy)
+}
+
+// orderings is a list of orderings which each specify
+// an ordered list of blocks (by number) to add to a single channel
+// and an ordered list of frames to read from the channel and submit
+// on L1. There will be one frame per block.
+// Depending on these lists, whether the channel is built as
+// as span batch channel, and whether the blocks are modified / invalidated
+// we expect a different progression of the safe head under Holocene
+// derivation rules, compared with pre Holocene.
+var orderings = []ordering{
+	{name: "case-0", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 3},       // regular case
+	{name: "case-1", blocks: []uint{1, 3, 2}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 1},       // out-of-order blocks
+	{name: "case-2", blocks: []uint{2, 1, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 0},       // out-of-order blocks
+	{name: "case-3", blocks: []uint{2, 2, 1, 3}, frames: []uint{0, 1, 2, 3}, safeHeadPreHolocene: 3, safeHeadHolocene: 0}, // duplicate block
+	{name: "case-4", blocks: []uint{1, 2, 3}, frames: []uint{2, 1, 0}, safeHeadPreHolocene: 3, safeHeadHolocene: 0},       // bad frame ordering
+	{name: "case-5", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 0, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 0},    // duplicate frames
+	{name: "case-6", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 0, safeHeadHolocene: 0,
+		isSpanBatch: true, blockModifiers: []actionsHelpers.BlockModifier{nil, blockFudger, nil}}, // partially invalid span batch (invalid payload)
+	{name: "case-7", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 1, safeHeadHolocene: 1,
+		isSpanBatch: false, blockModifiers: []actionsHelpers.BlockModifier{nil, blockSpudger, nil}}, // partially invalid singular batch channel (invalid batch)
+	{name: "case-8", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 3,
+		isSpanBatch: true, blockModifiers: []actionsHelpers.BlockModifier{nil, blockSpudger, nil}}, // partially invalid span batch (invalid batch?)
```

Comment:
So far these cases are failing to exercise any different behaviour under Holocene.

Reply:
The invalid payload handling isn't implemented yet. This doesn't happen at the batch stage, but later in the engine queue. Do we have tests that invalidate the _batch_ not the payload?

Reply:
As discussed on our call today, I"m going to look into adding such tests next.

Reply:
These have now been added.